language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -781,30 +781,64 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
             }
 
             // TODO: Handle type actuals for generics
-            Bytecode::BorrowGlobal(idx, _) => {
+            Bytecode::MutBorrowGlobal(idx, _) => {
                 let struct_definition = self.module().struct_def_at(*idx);
                 if !StructDefinitionView::new(self.module(), struct_definition)
                     .is_nominal_resource()
                 {
                     return Err(VMStaticViolation::BorrowGlobalNoResourceError(offset));
-                } else if !state.global(*idx).is_empty() {
+                }
+                if !state.global(*idx).is_empty() {
                     return Err(VMStaticViolation::GlobalReferenceError(offset));
                 }
 
                 let operand = self.stack.pop().unwrap();
-                if operand.signature == SignatureToken::Address {
-                    let nonce = self.get_nonce(&mut state);
-                    state.borrow_from_global_value(*idx, nonce.clone());
-                    self.stack.push(StackAbstractValue {
-                        signature: SignatureToken::MutableReference(Box::new(
-                            SignatureToken::Struct(struct_definition.struct_handle, vec![]),
-                        )),
-                        value: AbstractValue::Reference(nonce),
-                    });
-                    Ok(())
-                } else {
-                    Err(VMStaticViolation::BorrowGlobalTypeMismatchError(offset))
+                if operand.signature != SignatureToken::Address {
+                    return Err(VMStaticViolation::BorrowGlobalTypeMismatchError(offset));
+                }
+
+                let nonce = self.get_nonce(&mut state);
+                state.borrow_from_global_value(*idx, nonce.clone());
+                self.stack.push(StackAbstractValue {
+                    signature: SignatureToken::MutableReference(Box::new(SignatureToken::Struct(
+                        struct_definition.struct_handle,
+                        vec![],
+                    ))),
+                    value: AbstractValue::Reference(nonce),
+                });
+                Ok(())
+            }
+            Bytecode::ImmBorrowGlobal(idx, _) => {
+                let struct_definition = self.module().struct_def_at(*idx);
+                if !StructDefinitionView::new(self.module(), struct_definition)
+                    .is_nominal_resource()
+                {
+                    return Err(VMStaticViolation::BorrowGlobalNoResourceError(offset));
                 }
+
+                let freeze_ok = state
```

Comment:
Not blocking but I did find the name of `fn freeze_ok` a bit confusing. The function checks if there're any mutable references stored in variables or on stack. Should we change it to something more straightforward, like `fn exists_mutable_references`?

Reply:
There needs to be some collapsing of names here, but I prefer succinct judgments that we utilize and compose throughout the rules. In other words `exists_mutable_references` is too imperative for me. 

I would prefer us to use terms like `readable`, `writable`, and `freezable` and then compose them.